pkg/storagedistribution/storagedistribution.go

```diff
@@ -0,0 +1,390 @@
+		)
+		for instancesPerZone = requestedInstancesPerZone; instancesPerZone > 0; instancesPerZone-- {
+			capacityPerNode = minCapacityPerZone / uint64(instancesPerZone)
+			printCandidates("Candidate", []cloudops.StorageDecisionMatrixRow{row}, instancesPerZone, capacityPerNode)
+
+			instStorage, err := instanceStorageForRow(row, capacityPerNode, 0)
+			if err != nil {
+				if err == cloudops.ErrStorageDistributionCandidateNotFound {
+					printCandidates("Candidate failed to satisfy requirements.",
+						[]cloudops.StorageDecisionMatrixRow{row}, instancesPerZone, capacityPerNode)
+					continue
+				}
+
+				logrus.Errorf("failed to get storage spec for instance: %v", err)
+				return nil, 0, cloudops.ErrStorageDistributionCandidateNotFound
+			}
+
+			// additional check so that we are not overprovisioning
+			for instancesPerZone > 1 && instStorage.DriveCapacityGiB*instancesPerZone > minCapacityPerZone {
+				if instStorage.DriveCapacityGiB*(instancesPerZone-1) >= minCapacityPerZone {
+					instancesPerZone--
+				} else {
+					break // no more trim down possible
+				}
+			}
+
+			return instStorage, instancesPerZone, nil
+		}
+
+		printCandidates("Candidate failed as instances per zone exhausted",
+			[]cloudops.StorageDecisionMatrixRow{row}, 0, capacityPerNode)
+
+	}
+
+	return nil, 0, cloudops.ErrStorageDistributionCandidateNotFound
+}
+
+func instanceStorageForRow(
+	row cloudops.StorageDecisionMatrixRow,
+	requiredCapacity uint64,
+	requiredDriveSize uint64,
+) (*cloudops.StoragePoolSpec, error) {
+	instStorage := &cloudops.StoragePoolSpec{
+		DriveType: row.DriveType,
+		IOPS:      row.IOPS,
+	}
+
+	if requiredDriveSize > 0 {
+		// user has asked for a drive size. We can only create/update drives
+		// in increments of this size
+		if requiredDriveSize >= row.MinSize && requiredDriveSize <= row.MaxSize {
+			instStorage.DriveCapacityGiB = requiredDriveSize
+			remainingCapacity := int(requiredCapacity)
+			for remainingCapacity > 0 {
+				remainingCapacity -= int(requiredDriveSize)
+				instStorage.DriveCount++
+			}
+			return instStorage, nil
+		}
+
+		logrus.Debugf("skipping row as requiredDriveSize: %d not in [%d,%d]",
+			requiredDriveSize, row.MinSize, row.MaxSize)
+	} else {
+		if requiredCapacity >= row.MinSize {
+			// start from row's min drives to distribute the requiredCapacity across the drives
+			for driveCount := row.InstanceMinDrives; driveCount <= row.InstanceMaxDrives; driveCount++ {
+				driveSize := requiredCapacity / driveCount
+
+				if driveSize > row.MaxSize || driveSize < row.MinSize {
+					// drive is outside row's [min, max] bounds
+					continue
+				}
+
+				instStorage.DriveCapacityGiB = driveSize
+				instStorage.DriveCount = driveCount
+				instStorage.IOPS = row.IOPS
+				break
+			}
+
+			if instStorage.DriveCount == 0 {
+				return nil, cloudops.ErrStorageDistributionCandidateNotFound
+			}
+
+		} else {
+			logrus.Debugf("required capacity: %d GiB is lower than row's min size: %d GiB",
```

Comment:
why not try the next candidate row /  reduce the instances per zone and retry the same row?